Feature/c ffi

IceTDrinker opened the pull request:
### Resolves:

closes https://github.com/zama-ai/concrete-core-internal/issues/98

### Description

Requires #20 to be merged first
We probably also want #17 to have parallel BSK generation

Prototype C FFI operating on views for computation engines.

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (apply the `aws_test` to the PR to launch the tests on AWS)
* [x] The draft release description has been updated
* [x] Check for breaking changes and add them to commit message following the conventional commit [specification][conventional-breaking]

<!--
### Requires: `<link_your_required_issue_here>`
-->

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

IceTDrinker commented:
- I put it as draft as tests need a do-over with unchecked versions and serialization checked.
- C tests need to run on github to check the FFI is working as intended
- Requires a release workflow for github actions as well
- Docs needs to be added

This is a first step in the review process to check we are on the same page API-wise, all comments welcome, I'll continue working on the items I just listed above

IceTDrinker commented on Cargo.toml:
```diff
@@ -17,5 +18,8 @@ concrete-core = {path="concrete-core"}
 concrete-csprng = {path="concrete-csprng"}
 concrete-commons = {path="concrete-commons"}
 
+[profile.release]
+lto = "fat"
+
```
not sure about that one but it should be better so that the FFI has the maximum of things inlined I would say?

IceTDrinker replied:
Ok I'll resolve this then !

IceTDrinker commented on concrete-ffi/README.md:
```diff
@@ -0,0 +1,37 @@
+- use all the unchecked versions of the required engines
+- return 0 if everything went well, 1 otherwise
+
+The `view_buffers` indicates that for engines taking input and output ciphertexts, raw pointers pointing to a large enough buffer of the data type being used (e.g. `u64`) can be passed instead. The FFI function wrapper will build the appropriate ciphertext views using these pointers as underlying memory and then pass these views to the engine function to execute.
```
this is outdated and will be updated

IceTDrinker replied:
removed, this will be found in the docs

IceTDrinker commented on concrete-ffi/rustfmt.toml:
```diff
@@ -0,0 +1 @@
+imports_granularity="Module"
```
not sure this is needed

IceTDrinker commented on concrete-ffi/src/backends/fftw/engines/destruction.rs:
```diff
@@ -0,0 +1,40 @@
+use crate::utils::*;
+
+#[no_mangle]
+pub unsafe extern "C" fn fftw_engine_destroy_lwe_bootstrap_key_u64(
```
naming needs updating

IceTDrinker replied:
done locally

IceTDrinker commented:
One thing that seems pretty bad: cbindgen enables all the features on dependency crates, this sounds like a recipe for disaster down the line -_-

IceTDrinker commented:
Latest push adds C FFI tests to cargo xtask test and updates the used AMI so that tests can build and run the C tests

IceTDrinker commented:
rebased on main and fix `multithread` -> `parallel` feature rename

IceTDrinker commented:
Added parallel engine for bsk generation, updated EC2 AMI to run concrete-ffi tests.

mayeul-zama commented on concrete-ffi/src/backends/default/engines/lwe_ciphertext_creation.rs:
```diff
@@ -0,0 +1,109 @@
+        let input = get_ref_checked(input).unwrap();
+        let input_container = std::slice::from_raw_parts(input, lwe_size);
+
+        let heap_allocated_lwe_ciphertext_view = Box::new(
```
Creating a view allocates which may be undesirable and may affect performance.
Having LweCiphertextView64 be an opaque type makes this allocation necessary.

A possible solution is to use a `#[repr(C)] extern C` type for views so that a user of the C FFI may create one on the stack.

This comment also applies to other types of view

IceTDrinker replied:
Yep agreed, and as discussed, that's something we went for as it seemed ok when we exchanged on the topic.

It should not be too much work to provide an entry point that does not require passing views to it. In fact we can re-use the existing view API but have a raw ptr one that create the views on the fly and passes pointers to this view API.

I'll prepare a version where both are available and we'll see what we go with for the first release

IceTDrinker replied:
issue on this can be found here: https://github.com/zama-ai/concrete-core-internal/issues/158

closing this conversation

IceTDrinker commented:
Fixed wrong file mode on a concrete-ffi helper script.

Added docs.

Added raw_ptr_buffers versions of the relevant operators. Tests have not yet been updated, will wait for an opinion on this from Agnès when she gets back.

IceTDrinker commented:
Added a skeleton of concrete-ffi build release, we need to agree on how it is released (triggers mainly) and if the release is created automatically or if we just want to generate artifacts .

For now it triggers on PR so that I can check which artifacts are built and continue iterating on the workflow afterwards.

IceTDrinker commented:
After discussing this morning with @agnesLeroy here's what we will do:

- manual releases, so the workflow is just used to generate the required artifacts which will for now be uploaded manually to the releases
- release both APIs (view buffers and raw ptr buffers)

This means for this PR I still need to:
- [ ] add back tests with raw ptrs
- [ ] update the main rust docs with some infos on raw ptr buffers

User docs is tracked here: https://github.com/zama-ai/concrete-core-internal/issues/157

IceTDrinker commented:
the latest push is to check the script for the automated build part works as I expect

IceTDrinker commented:
This latest push disables the build on PR (it was successful here: https://github.com/zama-ai/concrete-core/actions/runs/2370873357)

Added tests for raw_ptr variants, updated the docs to give infos on raw_ptr_buffers variants

Based on https://github.com/zama-ai/concrete-core/pull/23 to fix the clippy lints issues

agnesLeroy commented:
> Another question, it seems most (if not all) the crates in the workspace are compatible with the 2021 edition of rust, concrete-ffi was created with 2021 edition by default by Cargo, @agnesLeroy should I open an issue to try and switch all our crates to 2021 ?

I'm not sure why we're not using the 2021 edition (I hope it's not a problem for compiler or top libs), but it's worth creating an issue, definitely! I'm checking with ThomasM if it could be an issue on their side.

IceTDrinker commented:
> > Another question, it seems most (if not all) the crates in the workspace are compatible with the 2021 edition of rust, concrete-ffi was created with 2021 edition by default by Cargo, @agnesLeroy should I open an issue to try and switch all our crates to 2021 ?
> 
> I'm not sure why we're not using the 2021 edition (I hope it's not a problem for compiler or top libs), but it's worth creating an issue, definitely! I'm checking with ThomasM if it could be an issue on their side.

I think it's mainly on the frontend/parser that it changes things IIRC rust is made to be compatible across language editions, so 2018 crates were able to have 2015 edition dependencies

IceTDrinker commented on .github/workflows/aws_tests.yml:
```diff
@@ -31,11 +31,11 @@ jobs:
           mode: start
           github-token: ${{ secrets.CONCRETE_ACTIONS_TOKEN }}
-          ec2-image-id: ami-016cd2805f00564a5
+          ec2-image-id: ami-0dae0bbbc42a4664a
```
so it's the same AMI as before but with CMake added, it is required to build the C tests

IceTDrinker replied:
detailed infos in our notion "CI infos" 🙂

IceTDrinker commented on .github/workflows/aws_tests.yml:
```diff
@@ -31,11 +31,11 @@ jobs:
           subnet-id: subnet-da319dd4
           security-group-id: sg-064a7184f24469c76
-          
+
```
probably yes

IceTDrinker commented on .github/workflows/ffi_github_release.yml:
```diff
@@ -0,0 +1,54 @@
+on:
+  # TODO: decide on how we trigger the release workflow
+  # workflow_dispatch:
+  pull_request:
```
Yes I will document the process on Notion, the latest PR only has the workflow dispatch enabled

IceTDrinker commented on concrete-ffi/README.md:
```diff
@@ -0,0 +1,10 @@
+This crate exposes an experimental C FFI to the `concrete-core` primitives. Using this FFI, any language can benefit from the FHE scheme proposed in `concrete`.
+
+This FFI is currently experimental and therefore unstable in terms of naming and exposed structures/entry points.
+
```
We can create the relevant sections in the Community already I think ? If we are missing issues on this I can create them, let me know !

IceTDrinker replied:
OK ! I have it in my TODO

IceTDrinker commented on concrete-ffi/src/backends/default/engines/glwe_ciphertext_creation.rs:
```diff
@@ -0,0 +1,138 @@
+
+/// Create a `GlweCiphertextView64`.
+///
+/// Refer to `concrete-core` implementation for detailed documentation.
```
Ah yes for sure ! Will do

agnesLeroy replied:
:cry: ok, it would have been nice but if it's not possible never mind.

IceTDrinker replied:
I'm not saying it's definitely impossible, it's just that the simple solution did not work 🙂 I'll try to see if there are ways to cross reference documentation in rust (hopefully there is!)

IceTDrinker replied:
It looks like it should be possible, maybe the --no-deps is not helping here, or maybe we need to have a crates.io dep, I'll check

IceTDrinker replied:
Issue here: https://github.com/zama-ai/concrete-core-internal/issues/171

From the test crate I found (see issue) it seems it would require "proper" crates.io dependencies

IceTDrinker commented on concrete-ffi/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+//! - catch panics (panics across an FFI boundary are Undefined Behavior) and print the panic
+//!   backtrace
+//! - check that every pointer is not null and well aligned
+//! - use all the checked versions of the required engines
```
true, it's a notable difference in that case

IceTDrinker replied:
modified to emphasize the destroy engine behavior.

Let me know if it's ok

IceTDrinker commented on concrete-ffi/src/backends/default/engines/lwe_ciphertext_discarding_addition.rs:
```diff
@@ -0,0 +1,125 @@
+///
+/// This function is [checked](crate#safety-checked-and-unchecked-functions).
+#[no_mangle]
+pub unsafe extern "C" fn default_engine_add_two_lwe_ciphertexts_u64_view_buffers(
```
oops ^^" that's a mix from the very ancient prototype and the engine names, will update

IceTDrinker replied:
updated the names of these functions

IceTDrinker commented on concrete-ffi/src/backends/default/entities/lwe_secret_key.rs:
```diff
@@ -0,0 +1,138 @@
+///
+/// This function is [checked](crate#safety-checked-and-unchecked-functions).
+#[no_mangle]
+pub unsafe extern "C" fn clone_lwe_secret_key_u64(
```
so this is potentially a bad reason but it's because we need to clone the LweSecretKey to be able to transmute them to GlweSecretKey, otherwise the original key gets consumed, it's specific to the compiler use case.

I can add support for cloning for other keys but we would not have relevant tests out of the box (the bootstrap test uses the clone + transmute of LweSecretKey to generate the required GlweSecretKey)

IceTDrinker commented on concrete-ffi/src/backends/default/engines/lwe_ciphertext_discarding_addition.rs:
```diff
@@ -0,0 +1,125 @@
+/// [Unchecked](crate#safety-checked-and-unchecked-functions) version of
+/// [`default_engine_add_two_lwe_ciphertexts_u64_view_buffers`]
+#[no_mangle]
+pub unsafe extern "C" fn default_engine_add_two_lwe_ciphertexts_unchecked_u64_view_buffers(
```
yep will be changed

IceTDrinker commented on concrete-ffi/src/lib.rs:
```diff
@@ -0,0 +1,124 @@
+//! - return 0 if everything went well, 1 otherwise and print the associated error message/panic
+//!   backtrace
+//!
+//! The `view_buffers` indicates that for engines taking input and output ciphertexts, ciphertext
```
agreed, I will update the section of the doc

agnesLeroy replied:
Looks great, thanks!

IceTDrinker commented on concrete-ffi/src/lib.rs:
```diff
@@ -0,0 +1,124 @@
+//! # Memory management
+//!
+//! All entities returned by the FFI are heap-allocated through `Box` and therefore require a call
+//! to the proper deallocation/destruction function to avoid memory leaks.
```
So a quick example in the README.md to show what people can do, and also the user docs for which we have the issue https://github.com/zama-ai/concrete-core-internal/issues/157

IceTDrinker replied:
Added an homomorphic addition example, but it's not small, it's a raw ptr version as well to be smaller 😕

agnesLeroy replied:
I've added some comment about it :wink: @damienligier we'll need your help for the example to check that it's cryptographically ok :wink: Actually it's based on the tests and we also need your help for them :smiley:

IceTDrinker replied:
The tests were taken from the prototype, so there may be room for improvement.

Otherwise I can try to re-use the examples from fixtures or from doc strings

agnesLeroy replied:
@IceTDrinker I really don't think it's a good idea at this stage to try and duplicate the statistical tests performed for concrete-core (that would take too much time, and I'm not even convinced it's actually useful). Maybe we should discuss this with @BourgerieQuentin actually :slightly_smiling_face:

BourgerieQuentin replied:
Of what I understand I guess it is not mandatory as this API is a kind of facade and that should not change the statistical tests, but it is up to you as you are responsible for this API. Actually I think a facade API should be only tested with a mock of the "backend" as the goal of this kind of API is to just to adapt an API.

IceTDrinker replied:
Hmm, what would the tests look like with such a mock backend ?

We want to check we are calling the right functions in the lib, and a mock backend kind of defeats that purpose ? 🤔

agnesLeroy replied:
The thing is, we can stick to the checks that the function calls are successful, but that won't ensure that the "correct" ones are called... I don't see how we could check this as of now, unless by performing some simple check on the result.

IceTDrinker commented on concrete-ffi/src/seeders/unix_seeder.rs:
```diff
@@ -0,0 +1,79 @@
+    catch_panic(|| {
+        check_ptr_is_non_null_and_aligned(result).unwrap();
+        let heap_allocated_seeder_builder = Box::new(SeederBuilder::new(Box::new(move || {
+            let seed = u128_from_two_u64(seed_high_64, seed_low_64);
```
ah 😅
u128 has no stable C equivalent hence why I created this in the first place. But yes for sure, if this is known to be bad I'll change it to something better

agnesLeroy replied:
I mean, the way you did it seems ok, but the fact that the seeder builder is for unique use is unsafe. Maybe we should make it clear in the name (to make it very clear it's only meant to be used once).

IceTDrinker replied:
actually re-reading the code, it's not as bad as I first thought as the dev random is used to generate seeds with the unix seeder, the issue of having the same secret remains though.

I'm trying to find a way to allow calling a closure only once, no real solution in sight for now

IceTDrinker replied:
haven't found a mechanism I like for this, static in closures don't work as all generated closures will re-use the same static -_-

I could store a generic state in it as it's boxed anyway to pass the FFI boundary, I'll explore that solution

IceTDrinker replied:
So the true downside now that I re-read this is that seeders share the same secret but will be as good/bad as /dev/random so in the end it may not be that big of a problem, who should we ping for seeders-related questions ?

agnesLeroy replied:
I would say @damienligier or JB? I've pinged Damien on this issue but he's super busy at the moment :slightly_frowning_face:

IceTDrinker replied:
As discussed this morning, trying to use a UnixSeederFactory more than once will panic with an error message stating that using it more than once is dangerous. Resolving this.

IceTDrinker commented on concrete-ffi/src/lib.rs:
```diff
@@ -0,0 +1,116 @@
+//! - return 0 if everything went well, 1 otherwise and print the associated error message/panic
+//!   backtrace
+//!
+//! Note that you can get more insights into the eventual panic using the `RUST_BACKTRACE` env
```
Updated wording

agnesLeroy commented on concrete-ffi/tests/test_lwe_bootstrap_w_bsk_serde.c:
```diff
@@ -0,0 +1,689 @@
+
+#include "utils.h"
+
+const int PRECISION = 4;
```
Maybe we'll want to test various precision levels, but we may be reworking those tests in the future so I'm not sure. Ideally we'd want to test the same parameter sets as in the fixtures. Also, this seems to be common to all tests, maybe we could avoid redefining it in each file?

agnesLeroy replied:
Also, maybe just like in the README, PRECISION could be renamed into MESSAGE_BITS?

IceTDrinker replied:
yes most likely, I re-used some already existing examples from the previous prototype

IceTDrinker replied:
done, the constants are now in utils.h and they were renamed in README.md

agnesLeroy commented on concrete-ffi/tests/test_lwe_bootstrap_w_bsk_serde.c:
```diff
@@ -0,0 +1,689 @@
+
+void encode_and_expand_lut(uint64_t *output, size_t output_size, size_t out_precision,
+                           const uint64_t *lut, size_t lut_size) {
+  size_t output_len = output_size;
```
Do we really need output_len?

IceTDrinker replied:
that could be a leftover from one of my internal refactorings, I'll clean it up

agnesLeroy commented on concrete-ffi/tests/test_lwe_bootstrap_w_bsk_serde.c:
```diff
@@ -0,0 +1,689 @@
+const int PRECISION = 4;
+const int SHIFT = 64 - (PRECISION + 1);
+
+void encode_and_expand_lut(uint64_t *output, size_t output_size, size_t out_precision,
```
Here a small explanation of what the content of the LUT is and what the bootstrap result will be would be nice :slightly_smiling_face:

IceTDrinker replied:
yes 🙂

agnesLeroy replied:
What you wrote seems clear to me, thanks! No sorry I was actually confused and thought we were building a "specific" LUT but I had misread the code. :thinking:

agnesLeroy commented on concrete-ffi/tests/test_lwe_bootstrap_w_bsk_serde.c:
```diff
@@ -0,0 +1,689 @@
+  double pbs_variance = 0.00000000000001;
+  double encryption_variance = 0.0000000001;
+  size_t glwe_dimension = 1;
+  size_t input_lwe_dimension = 2;
```
Does this value for the LWE dimension actually yields a 4-bits precision?

agnesLeroy commented on concrete-ffi/tests/test_lwe_bootstrap_w_bsk_serde.c:
```diff
@@ -0,0 +1,689 @@
+  printf("Expected: %f, Obtained: %f\n", expected, obtained);
+  double abs_diff = abs(obtained - expected);
+  double rel_error = abs_diff / fmax(expected, obtained);
+  assert(rel_error < 0.01);
```
Why this value for the relative error? Actually, do we really want to add such a verification here? Isn't it enough to just test that the engines are called successfully, and rely on core's tests? I remember we had a discussion about this, and that we said we could keep those tests for now... but seeing the amount of code it represents and the questions about the parameter sets and the checks themselves, I'm wondering if this is a good idea. Maybe we could add a disclaimer in the tests to state that they are not meant to be exhaustive, nor very accurate statistically speaking, since we already have concrete-core's tests, but that we still do some "minor" verification just to be sure the C API behaves correctly?

IceTDrinker replied:
Yes the idea was to have some form of checking that the C API is behaving mostly how we want it to behave, building on the trust from the existing tests for `concrete-core`

These tests were converted from the previous prototype, so I did not put a lot of thought into them :|

IceTDrinker replied:
Follow-up issues on these tests: https://github.com/zama-ai/concrete-core-internal/issues/172

agnesLeroy commented on concrete-ffi/src/backends/default/engines/destruction.rs:
```diff
@@ -0,0 +1,342 @@
+
+        // Here we use the unchecked version to process the memory of the underlying boxed object
+        // Otherwise we may be moving out of the box through a clone and the original memory may not
+        // be properly zeroed out.
```
Sounds good to me

agnesLeroy commented on concrete-ffi/src/backends/default/entities/lwe_keyswitch_key.rs:
```diff
@@ -0,0 +1,95 @@
+/// This function is [checked](crate#safety-checked-and-unchecked-functions).
+#[no_mangle]
+#[cfg(feature = "serde_serialize")]
+pub unsafe extern "C" fn serialize_lwe_keyswitching_key_u64(
```
Yes! Could you open an issue for it so we don't forget?

agnesLeroy commented on concrete-ffi/README.md:
```diff
@@ -0,0 +1,111 @@
+
+DISCLAIMER: the parameters in the example below are insecure and for example purposes only.
+
+Here is a small-ish example of how to call `concrete-core` from C through `concrete-ffi` to compute the homomorphic addition of two ciphertexts. This needs to be linked agains `libconcrete_ffi`
```
against

agnesLeroy commented on concrete-ffi/README.md:
```diff
@@ -0,0 +1,111 @@
+#include <stdlib.h>
+#include <tgmath.h>
+
+const int PRECISION = 4;
```
Maybe this should be renamed into MESSAGE_BITS?

agnesLeroy commented on concrete-ffi/README.md:
```diff
@@ -0,0 +1,111 @@
+    printf("Comparing output. Expected %f, Obtained %f\n", expected, obtained);
+    double abs_diff = abs(obtained - expected);
+    double rel_error = abs_diff / fmax(expected, obtained);
+    assert(rel_error < 0.002);
```
Here why is it 0.002? Why are the expected/obtained variables cast to double? Could we keep them as integers and compare them exactly? @damienLigier am I wrong?

IceTDrinker replied:
converted previous tests, open to suggestions

agnesLeroy commented on concrete-ffi/README.md:
```diff
@@ -0,0 +1,111 @@
+
+This FFI is currently experimental and therefore unstable in terms of naming and exposed structures/entry points.
+
+## A small example
```
Maybe just "An example"?

mayeul-zama reviewed (approved):
The C API looks great, thank Arthur!

Do you want the `concrete-ffi` crate to live outside of this repo eventually?
If so, merging this PR would leave "temporary" code in this repo which may be undesirable

Should we rather use the checked version of the API in the compiler?

IceTDrinker commented on concrete-ffi/tests/test_lwe_encrypt_mul_cleartext_decrypt.c:
```diff
@@ -0,0 +1,279 @@
+  double variance = 0.000000001;
+
+  // We generate the key
+  size_t lwe_dimension = {10};
```
No idea why those are here 🤣 I use clang-format and somehow this must have had a weird format, I'll change that !

IceTDrinker replied:
done

IceTDrinker commented on concrete-ffi/tests/test_lwe_bootstrap_w_bsk_serde.c:
```diff
@@ -0,0 +1,688 @@
+      default_engine, output_lwe_dimension, &output_lwe_sk);
+  assert(lwe_out_key_ok == 0);
+
+  LweSecretKey64 *output_lwe_sk_clone = NULL;
+  int lwe_out_sk_clone_ok = clone_lwe_secret_key_u64(output_lwe_sk, &output_lwe_sk_clone);
+  assert(lwe_out_sk_clone_ok == 0);
+
+  GlweSecretKey64 *output_glwe_sk = NULL;
+  int glwe_sk_ok = default_engine_transmute_lwe_secret_key_to_glwe_secret_key_u64(
+      default_engine, &output_lwe_sk_clone, poly_size, &output_glwe_sk);
+  assert(glwe_sk_ok == 0);
+  assert(output_lwe_sk_clone == NULL);
```
yep understood, I can create that in utils.c

IceTDrinker replied:
done

mayeul-zama reviewed (commented):
To reduce code duplication, I think `unchecked` could be a compilation flag

IceTDrinker commented:
> To reduce code duplication, I think `unchecked` could be a compilation flag

Hmm, I'm not sure it would be easy to write the code unless the get_ref_checked/get_mut_checked have their behavior change depending on the flag, which I'm not necessarily a fan of

IceTDrinker commented on concrete-ffi/tests/utils.c:
```diff
@@ -0,0 +1,67 @@
+  if (unix_seeder_available) {
+    // DANGER DANGER DANGER DANGER DANGER DANGER
+    // HIGHLY UNSAFE ONLY FOR TESTING PURPOSES
+    uint64_t secret_high_64 = 0;
+    uint64_t secret_low_64 = 0;
```
as discussed this question has been asked to people more knowledgeable on secret generation as we don't yet have a solution for this, I'll resolve for now and this PR will be updated (or not depending on how much work is required to bring said secret generation to code)

BourgerieQuentin commented on concrete-ffi/tests/test_lwe_encrypt_mul_cleartext_decrypt.c:
```diff
@@ -0,0 +1,279 @@
+  assert(sk_ok == 0);
+
+  // We generate the texts
+  uint64_t *input_ct_buffer = aligned_alloc(U64_ALIGNMENT, sizeof(uint64_t) * (lwe_dimension + 1));
```
Yes by default it call malloc, if they are requirement on alignment we just need to known what they are those requirements.

IceTDrinker replied:
There are constants in the concrete-ffi.h you can use to have aligned allocs.

For u64 and u32 it should be fine (as native types should have the same aligment between C++ and Rust), we just exposed the alignment constant just in case.

agnesLeroy commented:
> The C API looks great, thank Arthur!
> 
> Do you want the `concrete-ffi` crate to live outside of this repo eventually? If so, merging this PR would leave "temporary" code in this repo which may be undesirable
> 
> Should we rather use the checked version of the API in the compiler?

@mayeul-zama regarding the repositories, for now we prefer to keep them in the same repository: it avoids having to setup another CI, and things are still pretty unstable so it seems more practical for us.
